### PR TITLE
ENH: add Py3.5-style matmul operator (e.g. A @ B) to sparse linear algebra

### PR DESCRIPTION
**Pull request by jakevdp**
Simple addition; fully backward compatible.

### CODE REVIEW

File: scipy/sparse/base.py [hunk 1]
```diff
@@ -386,6 +386,16 @@ def __rmul__(self, other):  # other * self
                 tr = np.asarray(other).transpose()
             return (self.transpose() * tr).transpose()
 
+    #####################################
+    # matmul (@) operator (Python 3.5+) #
+    #####################################
+
+    def __matmul__(self, other):
+        return self * other
```
Comment (pv) [inline scipy/sparse/base.py]: Probably better as `self.__mul__(other)` so that NotImplemented is in correct layer.

**Comment by jakevdp** [comment]
Thanks @pv – I changed ``self * other`` to ``self.__matmul__(other)``.  Also made this change within ``dot()``, because I suppose it makes sense there as well.

**Comment by jakevdp** [comment]
Actually, now that I think about it, maybe it would be clearer in this case just to do
```
__matmul__ = __mul__
__rmatmul__ = __rmul__
```

File: scipy/sparse/base.py [hunk 2]
```diff
@@ -386,6 +386,16 @@ def __rmul__(self, other):  # other * self
                 tr = np.asarray(other).transpose()
             return (self.transpose() * tr).transpose()
 
+    #####################################
+    # matmul (@) operator (Python 3.5+) #
+    #####################################
+
+    def __matmul__(self, other):
+        return self.__matmul__(other)
```
Comment (larsmans) [inline scipy/sparse/base.py]: I suppose you meant `__mul__` for the implementation. this is an infinite regress.

File: scipy/sparse/base.py [hunk 3]
```diff
@@ -386,6 +386,16 @@ def __rmul__(self, other):  # other * self
...
+
+    def __rmatmul__(self, other):
+        return self.__rmatmul__(other)
```
Comment (larsmans) [inline scipy/sparse/base.py]: Same here.

File: scipy/sparse/base.py (hunk 2 above)
Comment (jakevdp) [inline scipy/sparse/base.py]: Right... oops!

**Comment by larsmans** [comment]
Fine with me. In fact, why don't we rename `__mul__` to `__matmul__` and define `__mul__ = __matmul__`?

File: scipy/sparse/base.py [hunk 4]
```diff
@@ -249,7 +249,7 @@ def dot(self, other):
         array([ 1, -3, -1], dtype=int64)
 
         """
-        return self * other
+        return self.__mul__(other)
```
Comment (pv) [inline scipy/sparse/base.py]: I think this is incorrect in principle --- `dot` should not return NotImplemented.

Comment (jakevdp) [inline scipy/sparse/base.py]: OK – I'm fine changing this back – I just thought it would be more consistent given your previous comment.

Comment (pv) [inline scipy/sparse/base.py]: The difference between the two is that `__mul__`, `__matmul__` correspond to Python's binop override mechanism, where `NotImplemented` as a return value has a special meaning, and will automatically result to TypeError if both sides punt.

`dot` on the other hand is an ordinary method, not in the binop system, which is why it should dispatch to `*`.

Comment (jakevdp) [inline scipy/sparse/base.py]: Ah – got it. Thanks!

**Comment by charris** [comment]
What is the return type?

**Comment by charris** [comment]
Other things to ponder

* `@` does not work with scalars, that is what `*` is for.
* If `v` is 1-D, `v@v` is a scalar.
* Eventually matrix needs to go away.

**Comment by jakevdp** [comment]
> What is the return type?

I would imagine the return type is whatever ``__mul__`` returns currently, as it basically implements matrix multiplication (and a matrix times a scalar is a subset of matrix multiplication).

**Comment by argriffing** [comment]
I haven't been following matmul much, but to clarify

> @ does not work with scalars

vs.

> (and a matrix times a scalar is a subset of matrix multiplication)

I searched 'scalar' in @njsmith's http://legacy.python.org/dev/peps/pep-0465/ and found this:

> 0d (scalar) inputs raise an error. Scalar * matrix multiplication is a mathematically and algorithmically distinct operation from matrix @ matrix multiplication, and is already covered by the elementwise * operator. Allowing scalar @ matrix would thus both require an unnecessary special case, and violate TOOWTDI.

and

> the consensus semantics for @ are that it should raise an error on scalars.

**Comment by jakevdp** [comment]
OK – that seems very strange to me. My guess is that the same docs would recommend against using ``*`` for matrix multiplication as is done here, if we want to be strict about it :smile: 

I think given the current everythin-goes syntax, it probably makes sense to allow ``@`` to work with scalars, but I'm willing to defer to whatever the consensus is.

**Comment by njsmith** [comment]
There is an argument -- I'm not sure if I buy it myself or not -- that sparse @ dense (and dot(sparse, dense), once that is implemented) should return an np.ndarray instead of an np.matrix. The argument is that we want to eventually get rid of matrix, and sparse's habit of returning matrix objects creates endless headaches for users (see e.g. scikit-learn's safe dot functions), so when adding new functionality we should take the opportunity to provide something closer to what users want.

The argument against is that this still won't really help much with getting rid of matrix so long as sparse * works the way it does, and fixing that will probably require providing a whole new sparse package, so in the mean time we should keep scipy.sparse internally consistent.

**Comment by jakevdp** [comment]
I'm fine with all of that. I just think the least confusing thing overall would be basically ``__matmul__ = __mul__``.

**Comment by njsmith** [comment]
And yes, the end goal I think we all agree on (and that the PEP focuses on) is that * should be elementwise and @ should be what it says there. If something else makes sense as a temporary transitional thing then *shrug*.

**Comment by jakevdp** [comment]
@njsmith – do you think that holds for linear operators as well? Mathematically, linear operators are simply a higher-level abstraction of the concept of multiplication. Artificially subdividing a unified mathematical concept based on syntax decisions made by a bunch of computer scientists seems overly pedantic to me.

**Comment by charris** [comment]
@jakevdp Well, the mathematically inclined did argue for scalars being diagonal matrices, but I don't think that is what you had in mind. One way to look at it is that scalar multiplication is part of the definition of a vector space. The definition makes sure that scalar multiplication behaves like a linear operator, but you don't need the idea of a linear operator to introduce vector spaces. Morphisms come a bit later ;) I think the distinction is between objects with dimensions, and those without.

**Comment by jakevdp** [comment]
Interesting piece of data – NumPy 1.10 allows scalars in ``dot``, but raises a ``ValueError`` with ``@``:

```
>>> np.dot(np.arange(4), 5)
array([ 0,  5, 10, 15])
>>> np.arange(4) @ 5
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-f3b7df22bf35> in <module>()
----> 1 np.arange(4) @ 5
```

**Comment by njsmith** [comment]
@jakevdp: The idea is that if you want to make an operation dispatch on the shape (which we do, b/c the automagic promotion/demotion thing for 1d inputs is super-handy), then you need to be rigorous about shapes. In NumPy there is a 100% standard way for scalars to be promoted to matrices: you broadcast them. Except that `dot` implicitly promotes them by multiplying by the identity matrix, which is totally different. Maybe think about it this way: if we wanted scalars to implicitly promote to the equivalent linear operators, then we could do that, but we'd have to change `+` too. (`mat + 1` would return a copy of mat with the diagonal incremented, but nothing else, because `(mat + 1) @ foo` would need to be equivalent to `mat @ foo + 1 @ foo`.)

**Comment by jakevdp** [comment]
Got it! Thanks for the explanation.

So do you think it's worth adding logic here to raise an error when ``__matmul__`` sees a scalar?

**Comment by charris** [comment]
It would be good if code written for numpy using `@` would work the same with sparse, and vice versa.

**Comment by jakevdp** [comment]
OK – now the functions raise a ValueError for scalar inputs.

File: scipy/sparse/linalg/interface.py [hunk 5]
```diff
@@ -367,6 +367,12 @@ def dot(self, x):
                 raise ValueError('expected 1-d or 2-d array or matrix, got %r'
                                  % x)
 
+   def  __matmul__(self, other):
```
Comment (charris) [inline scipy/sparse/linalg/interface.py]: Indentation error.

Comment (jakevdp) [inline scipy/sparse/linalg/interface.py]: Thanks

File: scipy/sparse/linalg/interface.py [hunk 6]
```diff
@@ -367,6 +367,12 @@ def dot(self, x):
                 raise ValueError('expected 1-d or 2-d array or matrix, got %r'
                                  % x)
 
+    def  __matmul__(self, other):
```
Comment (charris) [inline scipy/sparse/linalg/interface.py]: I think you need one less space after `def`.

File: scipy/sparse/linalg/interface.py [hunk 7]
```diff
@@ -367,6 +367,12 @@ def dot(self, x):
                 raise ValueError('expected 1-d or 2-d array or matrix, got %r'
                                  % x)
 
+    def __matmul__(self, other):
+        if np.isscalar(other):
+            raise ValueError("Scalar operands are not allowed, "
+                             "use '*' instead")
+        return self.__mul__(other)
+
     def __rmul__(self, x):
```
Comment (pv) [inline scipy/sparse/linalg/interface.py]: `__rmatmul__`?
The subclasses probably define implementations that do something.

Comment (jakevdp) [inline scipy/sparse/linalg/interface.py]: Sure – I initially left it out because there's no right-side equivalent of ``dot()``, but I can add it now.

**Comment by rgommers** [comment]
This should go into 0.17.0. All comments seem to be addressed. Any objections to merging this as is?

**Comment by jakevdp** [comment]
I think it's ready to go

**Comment by perimosocordiae** [comment]
+1 here, LGTM.

**Comment by pv** [comment]
No tests, though.

**Comment by jakevdp** [comment]
How would one write tests for this that would not break earlier Python versions? Use exec?

**Comment by njsmith** [comment]
Probably eval instead of exec (b/c its syntax is the same between 2 and 3),
but yeah, basically.

**Comment by ev-br** [comment]
Or `from scipy._lib.six import exec_`

**Comment by pv** [comment]
Or using `operator.matmul`

**Comment by ev-br** [comment]
And this should probably be mentioned in the release notes, too.

**Comment by jakevdp** [comment]
Added some tests. Didn't run them locally, so Travis will tell us how good my mental Python interpreter is :)

**Comment by jakevdp** [comment]
The only thing failing now is the PEP8 test. I can't figure out what's going on... it says there are too many blank lines (3) at line 59, but there is only [one blank line there](https://github.com/jakevdp/scipy/blob/8b2dc12a222b3da70f9da696fe90348eab5779bf/scipy/sparse/tests/test_base.py#L59). I'm unable to reproduce this error locally (though I can make the same message occur by adding two extra lines there). Any ideas?

**Comment by argriffing** [comment]
Maybe it's a pep8 checker bug.

**Comment by charris** [comment]
I can't see a  problem either.

**Comment by pv** [comment]
The issue is that Travis-CI runs the tests after merging to master --- the merge produces blank lines.
Can be fixed by rebasing this PR and fixing up afterward (which also would also clean up the history).

**Comment by jakevdp** [comment]
@pv – thanks.

Any issues that need addressing before this is merged?

**Comment by rgommers** [comment]
@jakevdp the suggestion was to rebase on master and squash all the small typo/fix commits before merging.

I've only skimmed the added tests, but they look fine to me.

**Comment by jakevdp** [comment]
@rgommers - Would you like me to rebase/squash and force-push to this branch? I thought @pv meant we'd do that at merge.

**Comment by rgommers** [comment]
yes please. i think that's also what @pv meant.

**Comment by jakevdp** [comment]
OK – squashed.

**Comment by rgommers** [comment]
All clean and green. Anyone who reviewed in more detail than me want to finalize this?

**Pull request merged**